cmd/rekor-cli/app/timestamp.go

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package app
+
+import (
+	"encoding/asn1"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/sassoftware/relic/lib/pkcs9"
+	"github.com/sigstore/rekor/cmd/rekor-cli/app/format"
+	"github.com/sigstore/rekor/pkg/generated/client/timestamp"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+type fileFlag struct {
+	value string
+}
+
+func (f *fileFlag) String() string {
+	return f.value
+}
+
+func (f *fileFlag) Set(s string) error {
+	if s == "" {
+		return errors.New("flag must be specified")
+	}
+	if _, err := os.Stat(filepath.Clean(s)); os.IsNotExist(err) {
+		return err
+	}
+	f.value = s
+	return nil
+}
+
+func (f *fileFlag) Type() string {
+	return "fileFlag"
+}
+
+func addTimestampFlags(cmd *cobra.Command) error {
+	cmd.Flags().Var(&fileFlag{}, "request", "path to an RFC 3161 timestamp request")
+
+	cmd.Flags().Var(&fileFlag{}, "file", "path to a file containing the message to timestamp")
+
+	cmd.Flags().String("out", "response.tsr", "path to a file to write response.")
+
+	// TODO: Add a flag to upload a pre-formed timestamp response to the log.
+	// TODO: Add a flag to indicate a JSON formatted timestamp request/response.
+	return nil
+}
+
+func validateTimestampFlags() error {
+	requestStr := viper.GetString("request")
+	fileStr := viper.GetString("file")
+	if requestStr == "" && fileStr == "" {
+		return errors.New("request or file must be specified")
+	}
+	return nil
+}
+
+func createRequestFromFlags() (*pkcs9.TimeStampReq, error) {
```

Comment:
Is there existing tooling to easily create the TSR... I'm thinking a better UX would be to support some additional command line options like https://www.openssl.org/docs/man1.0.2/man1/ts.html instead of requiring a pre-formed binary input.

Reply:
TSQ :P

The golang tools I've seen that relied on https://github.com/digitorus/timestamp. They implement everything openssl provides besides the random seed and specifying the digest bytes.

(We can't use their CreateResponse since we don't implement a crypto signer)

It would be pretty easy to do add support for these options myself.

Reply:
OK added nonce bool to specify a nonce, string for TSA policy, string to specify the digest hash.

the only one from openssl i didn't include was the cert bool -- rekor always adds the cert chain. i could add it anyway though.